### ISSUE TITLE
Running WinForms apps in x64 mode on macOS

### ISSUE THREAD

**Issue by CNAmira** [mono/mono#6701]
According to the release note of Xcode 9.3, macOS will remove the support of running 32-bit application without compromise in future version, and they include a 'no32exec' mode in the beta version of macOS 10.13.4 which 32-bit applications will no longer run on.
System.Windows.Forms haven't been port to Cocoa API, so all WinForms apps won't runs on future macOS.
Recently I referenced to XplatUICarbon and managed to make a Cocoa version of XplatUI, but I haven't understand how it works, it would be better if anyone who can give me some thoughts.

**Comment by akoeplinger** [comment]
Correct, at the moment we don't support WinForms on x64 macOS. We'd really appreciate a pull request making this work.

**Comment by akoeplinger** [comment]
@CNAmira you said you managed to make a Cocoa version of XplatUI, do you mean something that was able to run a simple WinForms Hello World? Is the code available somewhere?

Even if it's just a first step it would be amazing to have it 😄

**Comment by CNAmira** [comment]
Not yet.

**Comment by migueldeicaza** [comment]
I looked into this yesterday

The good news is that Carbon now works on 64 bits.  I have long assumed Carbon died with 32, so either my memory is faulty or Apple at some point reversed course and added 64 bit support to it.  

Anyways, we now have a path to 64 but support.   

What needs to be done is to create a 64bit backend.   It should probably be a replica of the 32 bit one, with some bits shared where possible.   

The 64 bit version needs to have most of the structures in Structs.xs upgraded to 64 bits.   I did a partial review last night and I got a good sense of what needs doing.   Will continue the audit tonight.

Beyond the 64bit structure update we probably need to upgrade many calls that we make into Carbon that were suitable or requires back in 2005/2006 (either due to platform requirements and back compact needs) and replace them with new ones.

I haven’t found replacement for all, but WebKit and an old fork From adobe of WebKit have enough clues about what to do here.

So those would be the two stages.

**Comment by CNAmira** [comment]
In my memory most Carbon functions on 64bits are just stubs.
I built a simple Carbon application in Xcode, it only worked when I switch the architecture to i386 other than x86_64.

**Comment by migueldeicaza** [comment]
we will have to wait until we find those, we only use a handful of Carbon apis.

An alternative is to build a new backend on top of Cocoa, about 5000 lines of code.   To do this, I would dynamically load Xamarin.Mac, but it would be an external dependency and we would need to figure out how to distribute this.

**Comment by 0xced** [comment]
> In my memory most Carbon functions on 64bits are just stubs.

@CNAmira The first error I encountered when running a WinForms app on 64bits was that `GetWindowPort` was not found:
```
System.EntryPointNotFoundException: GetWindowPort
  at (wrapper managed-to-native) System.Drawing.MacSupport.GetWindowPort(intptr)
  at System.Drawing.MacSupport.GetCGContextForView (System.IntPtr handle) [0x0007f] in <fff295db00f34819a42b2efd9e41c64c>:0 
  at System.Drawing.Graphics.FromHwnd (System.IntPtr hwnd) [0x00042] in <fff295db00f34819a42b2efd9e41c64c>:0 
  at System.Windows.Forms.Control.CreateGraphics () [0x00019] in <2c6cbf9dc9fa4171bf0a705a710f3831>:0 
  at (wrapper remoting-invoke-with-check) System.Windows.Forms.Control.CreateGraphics()
```
We can see that `GetWindowPort` is external (uppercase T) in 32bits but is non-external on 64bits (lowercase t):
```
$ nm --arch=i386 /System/Library/Frameworks/Carbon.framework/Versions/Current/Frameworks/HIToolbox.framework/HIToolbox | grep GetWindowPort$
002b2eb1 T _GetWindowPort
$ nm --arch=x86_64 /System/Library/Frameworks/Carbon.framework/Versions/Current/Frameworks/HIToolbox.framework/HIToolbox | grep GetWindowPort$
00000000002386eb t _GetWindowPort
```

I don't know how `DllImport` is implemented on macOS, but I’m not surprised that it can't bind a non-external symbol.

So, it’s not a strictly a stub, but it has been changed to be inaccessible from outside.

<hr>

> An alternative is to build a new backend on top of Cocoa, about 5000 lines of code. To do this, I would dynamically load Xamarin.Mac, but it would be an external dependency and we would need to figure out how to distribute this.

@migueldeicaza Before creating `XplatUICocoa.cs` and blindly starting to implement it I thought I’d check GitHub to see someone already started to work on it. Do you already have a prototype of a Cocoa backend or are these 5000 lines of code you are talking about just an estimation?

Also, I really think the way forward is to use Cocoa, not Carbon. Qt for example took that direction in 2009 if I remember correctly.

**Comment by filipnavara** [comment]
@0xced We have a fork (https://github.com/filipnavara/mac-playground) that has XplatUICocoa mostly implemented. However it's not matter of simple cut & paste to plug it in back to Mono. We had some design choices early on, such as using the System.Drawing implementation on top of Cocoa, that make it complicated to use it as drop-in binary replacement. It works somewhat as a source-level replacement where applications link to our System.Window.Forms and System.Drawing libraries.

That said, large parts of the code were back-ported to official repositories in one way or another. The System.Drawing changes were backported for the Cocoa version (https://github.com/mono/sysdrawing-coregraphics). I have a pending patch to fix System.Drawing on 64-bit Cocoa, but it depends on patches to libgdiplus that were not merged yet. And the non-Cocoa changes to SWF were 90% back-ported to official Mono repository.

The next step is taking the XplatUICocoa from our repository and making it a separate library that can be loaded as external driver from SWF (like https://github.com/migueldeicaza/CocoaDriver). I have done this locally as a proof of concept, so I know it's possible, but my effort was largely unfinished. There was still some glue code missing to support creating Graphics objects for non-client and client parts of the controls, so the rendering of controls was not perfect.

**Comment by PreferLinux** [comment]
I've just been struggling with this issue, because of problems compiling libgdiplus (and dependencies!) as 32-bit. Either way, I gave up on that, and tried [XQuartz](https://www.xquartz.org/). It seems to work OK in my specific application after a very brief test. For anyone else wanting to try it, here are the steps I followed (I already had Mono compiled and installed as 64-bit):
1. Install XQuartz. It puts stuff in /opt/X11
2. Compile and install libgdiplus as 64-bit _with_ X11. When running configure I needed to specify where it could find X11, which resulted in running
`CPPFLAGS=-I/opt/X11/include LDFLAGS=-L/opt/X11/lib ./configure`
3. Run the application, forcing the use of XplatUIX11. I needed to specify the X11 library path too, so I ran it with
`LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/X11/lib MONO_MWF_MAC_FORCE_X11=true mono ___.exe`

It's not necessarily perfect, but I hope it might be useful if anyone else has problems before a suitable driver is finished.

**Comment by tyronx** [comment]
Can we expect 64 bit winforms to be available at some point? 

I have a fairly memory hungry game in C# and thus, downgrading to 32 bit is not an option for me. Windows.Forms contains some extremely useful tools such as querying the clip board and screen size. The lack of winforms is one major reason why my macos support is so wonky.

I'd be happy to offer a bounty of 100 us dollars if someone manages to get a pull request merged into mono that adds 64 bit winforms support

**Comment by CNAmira** [comment]
Since Microsoft has released the full sources of WinForms and WPF, is it possibly to combine those sources and Wine's 'native' libs to provide a better WinForms experience?

**Comment by PreferLinux** [comment]
Assuming it is possible, it'd probably be no better than running .Net on Wine, which does work but not particularly well.

**Comment by migueldeicaza** [comment]
Exactly, that is a different solution, it would not help in this scenario.

The baseline is still there.   That said, since this was originally started, we did merge a lot of the work from @filipnavara which should get us on a path to have a Cocoa backend, but I have not checked recently on how much was missing or was completed.

**Comment by zmast** [comment]
What's the current status on this issue?
I was interested in fixing couple of issues in WF but the Carbon API is so deprecated I can't even find any documentation and knowing Apple I have a feeling they will soon knock it off completely.

The work from @filipnavara sounds very promising. It's not clear to me if it can be merged though.

**Comment by joyinfo** [comment]
The biggest issue I use the mono on macOS is visibility and z-order issue (It is also said in https://semisignal.com/mono-winforms-on-os-x/).

I tried mac-playground from @filipnavara. It works great. The visibility and z-order issues disappear. Also I can use many native components on macOS.

**Comment by thijs30** [comment]
@PreferLinux Did you get 64-bits compiling of Windows.Forms working with Xquartz?

**Comment by PreferLinux** [comment]
Yes, I did. Just make sure libgdiplus is compiled with X11 (https://github.com/mono/libgdiplus#build-instructions), and run the application with `LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/X11/lib MONO_MWF_MAC_FORCE_X11=true mono ___.exe`

In my use-case it seems to work reliably, but is somewhat slow. Obviously, if XQuartz isn't running, it has to be started when starting the application – in my experience that takes at least 5 – 10 seconds. Drawing in general seems pretty slow as well – e.g. owner-drawn ListBox. And after installing XQuartz it requires a log out and back in before it'll get used, which could make application installation more awkward.

But the main thing is, it works.

**Comment by thijs30** [comment]
@PreferLinux Well, I’m probably not as experienced as you are on this topic so don’t understand really how to do the things you said. 

Will this solution also work when building and compiling within Visual Studio?

**Comment by PreferLinux** [comment]
Basically you have to install libgdiplus (a native component of Mono used by System.Drawing, a substitute for Window's GDI+) from source, so that it is compiled with the correct options. And to use that you probably have to install the rest of Mono from source too, with the option `--with-libgdiplus=sibling` so that it finds the correct libgdiplus. It probably also requires installing Cairo from source so that it can use X11 too (`PKG_CONFIG_PATH="/opt/X11/lib/pkgconfig" ./configure --with-x --enable-xlib=yes --enable-xcb=no`), and possibly Pango as well (no special options needed) because it uses Cairo; and that would probably need to be done before compiling libgdiplus... So yeah, it is a bit complicated, and I can't recall what is actually required! (I distribute Mono and its dependencies with my application, all compiled from source; so while I've got it all sorted now I don't remember the exact details of what I had to compile myself before I was simply compiling _everything_ myself!)

The application itself can be compiled in Visual Studio the same as for Mono on Linux, which may well be exactly the same as for .Net on Windows.

Then when you run the application from Terminal, instead of just doing `mono whatever.exe` you need to run it like I posted above – which basically sets the environment variables `LD_LIBRARY_PATH` (so that libgdiplus can find XQuartz's libraries) and `MONO_MWF_MAC_FORCE_X11` (so that Mono uses X11 rather than native macOS), and also does the `mono whatever.exe` part.

**Comment by alexchandel** [comment]
@akoeplinger Do you know where this landed re. patches to libgdiplus, System.Drawing, and System.Windows.Forms?

 @filipnavara Was XplatUICocoa ever merged

**Comment by akoeplinger** [comment]
@alexchandel as far as I know the current situation is still the same: Mono's WinForms doesn't work on 64bit macOS.

**Comment by Sappharad** [comment]
Coincidentally I just tested this in the current version of visual studio for Mac yesterday. Yeah, you get a warning about the carbon API not being 64-bit and WinForms apps don’t run. 
I’m using a manually built copy of the port linked earlier in this thread.

**Comment by alexchandel** [comment]
@akoeplinger But were the libgdiplus patches merged upstream?

And does Mono now compile libgdiplus with the right options for this?  If not, is it a simple change to do this?

I'm speaking in reference to @filipnavara's comments:

> That said, large parts of the code were back-ported to official repositories in one way or another. The System.Drawing changes were backported for the Cocoa version (https://github.com/mono/sysdrawing-coregraphics). I have a pending patch to fix System.Drawing on 64-bit Cocoa, but it depends on patches to libgdiplus that were not merged yet. And the non-Cocoa changes to SWF were 90% back-ported to official Mono repository.

And is Mono preparing to adopt the Cocoa System.Drawing fork?

**Comment by akoeplinger** [comment]
@alexchandel libgdiplus is maintained by us so we are the upstream in this case and we ship the latest version :)

I'm not sure how much work is still left to adopt @filipnavara's work or complete the Cocoa backend.

**Comment by akoeplinger** [comment]
It already does, at least the ones from www.mono-project.com.

However there seems to be a problem since I tried and it crashes when trying to find `GdipCreateFromXDrawable_linux`:

```
$ DISPLAY=:0 LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/X11/lib MONO_MWF_MAC_FORCE_X11=true mono helloworld.exe
Unhandled Exception:
System.EntryPointNotFoundException: GdipCreateFromXDrawable_linux assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) System.Drawing.GDIPlus.GdipCreateFromXDrawable_linux(intptr,intptr,intptr&)
  at System.Drawing.Graphics.FromXDrawable (System.IntPtr drawable, System.IntPtr display) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/System.Drawing/System.Drawing/Graphics.cs:1790
  at System.Drawing.Graphics.FromHwnd (System.IntPtr hwnd) [0x000ef] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/System.Drawing/System.Drawing/Graphics.cs:1748
  at System.Windows.Forms.XplatUIX11.GetAutoScaleSize (System.Drawing.Font font) [0x00010] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/System.Windows.Forms/System.Windows.Forms/XplatUIX11.cs:3890
  at System.Windows.Forms.XplatUI.GetAutoScaleSize (System.Drawing.Font font) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/System.Windows.Forms/System.Windows.Forms/XplatUI.cs:677
  at System.Windows.Forms.Form.GetAutoScaleSize (System.Drawing.Font font) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/System.Windows.Forms/System.Windows.Forms/Form.cs:1531
  at System.Windows.Forms.Form..ctor () [0x00018] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/System.Windows.Forms/System.Windows.Forms/Form.cs:353
  at HelloWorld..ctor () [0x00000] in <f9a7b878c7c84b6ab16bdf7aa6b87fdd>:0
  at (wrapper remoting-invoke-with-check) HelloWorld..ctor()
  at HelloWorld.Main () [0x00001] in <f9a7b878c7c84b6ab16bdf7aa6b87fdd>:0
```

This is because we're building the Cairo library without X support on macOS: https://github.com/mono/bockbuild/blob/5b5e5d686742b38e089416afcfa5be0003db1157/packages/cairo.py#L24-L30

@filipnavara do you happen to know whether enabling that would cause any issues?

**Comment by filipnavara** [comment]
There is no easy way to build libgdiplus with X11 support on macOS. You have to build many of dependent libraries yourself (cairo, pango, harfbuzz at least) without sourcing them from Homebrew. I know the Mono builds may be doing that anyway but I tried doing it outside of Bockbuild and it was pain in the ass.

**Comment by git-hub-9999** [comment]
Just wondering if there is any hope for a solution to this?

If anyone knows how to resolve this, please contact me as I would be happy to pay for a solution (if that would expedite completion).

**Comment by zukzuk2020** [comment]
I would cheap in if somebody can help with this please. 

Would building mono 2.10.5 or 2.10.9 as a 64 bit package which was the last builds I believe with X11 support work and be simple enough as a workaround solution?

**Comment by glopesdev** [comment]
@zukzuk2020 @filipnavara unfortunately mac-playground is really not sustainable as it only covers the base WinForms assembly. Worse still is that it's inoperable with cross-dependencies (i.e. if your project references another dependency which references `System.Drawing` or anything like that, you would have to rebuild everything from scratch).

Any news on this front? Is anything expected to change with the move to .NET 5? Or are all WinForms projects simply dead on MacOS?

**Comment by filipnavara** [comment]
> @zukzuk2020 @filipnavara unfortunately mac-playground is really not sustainable as it only covers the base WinForms assembly. Worse still is that it's inoperable with cross-dependencies (i.e. if your project references another dependency which references `System.Drawing` or anything like that, you would have to rebuild everything from scratch).

That's not entirely accurate. It is possible to use 3rd-party code with the included System.Drawing library but it's hell to manage in the build system. You basically have to convince the build system to prefer certain assets over others. It is even possible to get the System.Drawing.Common facade over System.Drawing from a NuGet with carefully crafted PackageReference. As long as the exported API is compatible it works at runtime. Definitely not for the faint of heart and not something I would recommend though.

> Any news on this front? Is anything expected to change with the move to .NET 5? Or are all WinForms projects simply dead on MacOS?

Unless someone takes action to revive it I would consider it dead. We continue to maintain our code internally and use it in our product but we have no resources to invest into to it aside from maintaining the GitHub mac-playground mirror (and even that one tracks the wrong internal branch now :-/).

**Comment by Sappharad** [comment]
For what it's worth, I've still been using it (mac-playground) in multiple projects. It's better than nothing at all, even though I have to roll custom builds of every single dependency that references WinForms & related in order for it to work. In general I've been using it for ports of open source code because writing a brand new UI isn't worth my time and otherwise there would be no Mac build at all.

I was hoping .NET 5 moving everything to Nuget would make the problem easier instead of harder, but it's a shame to know I'll have continue with a complicated work-around.

I'm not able to help either, but just wanted to show my support for even some kind of basic not-very-good support like we had before x64 was a requirement.

**Comment by glopesdev** [comment]
@filipnavara @Sappharad I agree the mac-playground is an incredibly useful contribution and in no way I wanted to convey the idea I am not grateful for it.

However, for better or worse, WinForms continues to have a future as part of .NET 5 on windows, and so making a sustainable solution is valuable work which will remain relevant for the foreseeable future. Is there a path forward to merge mac-playground contributions back to Mono? Are contributions being accepted still with .NET Core and .NET Fx merger?

I just wanted to understand a bit better what are the hurdles involved in integrating it with the current Mono codebase. It looks like there might be a window of opportunity for doing this in time for .NET 5.

**Comment by ErMaqui** [comment]
I cannot understand why the fork of mac-playground cannot be merged here.
.NET are a great platform, and project mono are something vital for it.

Today, we are on 2020, not on 2002, when .net first release go to public and Mac OS and Linux have low market share. Now, yes, windows have a lot of market, but Mac OS have their users and Linux too. Even Microsoft are moving .NET to a "full cross-platform" (with the next .NET 5 and leaving Framework support), but, designer and frontend are windows only. Does that have some sense? "Ey, you can use .NET to do your app for something, but only for console apps if you want to use cross-platform. If you want to use it on any system, please search some different."

I cannot understand why.
And, I can understand Mono project members doesn't have time to refactor more than 5000 lines of source, but, why not Microsoft directly?

Today you can use VS on mac/linux. Today you can have SQLServer on linux. But you cannot compile a simple hello world with graphics on a mac.

**Comment by Sappharad** [comment]
> "Ey, you can use .NET to do your app for something, but only for console apps if you want to use cross-platform. If you want to use it on any system, please search some different."
> Today you can use VS on mac/linux. Today you can have SQLServer on linux. But you cannot compile a simple hello world with graphics on a mac.

Your statements imply that you don't understand the process of developing .NET Apps for Mac then. You can already do these things, you just need to use the native UI tools and classes provided by Xamarin and included with Visual Studio. You can build fully native graphical applications, you just need to build them using native tools.

This thread is about using **Windows** Forms, which was created for native Windows UI's on Windows. The people who want it on macOS (including me) just want it for ease of porting so I don't need to re-create a massive open source app that gets changes faster than I have time to keep up with myself.

If someone is building a graphical application from scratch and they want to support multiple platforms, they design for that from the start by using something like Eto Framework (comparable to using WinForms) or Avalonia (comparable to WPF).

**Comment by iMonZ** [comment]
Is something new here?

**Comment by glopesdev** [comment]
Doesn't look like it.

It's interesting, though, to see the flurry of bug reports referencing this issue, which is indicative of how widespread the use of WinForms remains.

**Comment by migueldeicaza** [comment]
A whole new backend needs to be written on top of Cocoa.   It is not impossible, but it requires some dedication.

Windows.Forms in Mono has not been in active development since about 2009, and the reason we discontinued its development at the time was that too many applications at the time were doing P/Invoke that we could not possibly emulate, or relied on internals of the WndProc system that were not completely emulated.   It certainly works for some scenarios, but not all.   At the time we discontinued this work at Novell, Silverlight looked more promising, and that is what we did.

It si possible to implement this, but it will take some effort, and someone committed to doing that work.   Getting something on the screen for someone familiar with Cocoa could probably take a couple of days, and a decent backend done in probably a month.

**Comment by filipnavara** [comment]
As stated above, we have done the backend work at https://github.com/emclient/mac-playground (somewhat buildable branch at https://github.com/emclient/mac-playground/tree/build-fixes). We do plan to keep maintaining it for foreseeable future and we have it running on .NET 6 already. However, we have no resources to spare to make the CI working or add NuGet builds to make it easy to consume.

**Comment by zukzuk2020** [comment]
filipnavara's https://github.com/emclient/mac-playground works great for us. Fingers crossed some new macOS update doesn't break anything there, but if this is merged with Mono, man... it would be a relief!

**Comment by filipnavara** [comment]
> Fingers crossed some new macOS update doesn't break anything there

We do get some small breakage all the time but we keep fixing it. I'll update the public repository once we finish our internal company testing.

**Comment by HinTak** [comment]
Stating the obvious - the large batch of linked reports in the middle of 2020 is likely because people were upgrading to Mac OS X 10.15, which also removes the ability to run applications in 32-bit mode. 10.15 was apparently out in oct 2019, but people were probably delaying upgrading until Big Sir was announced in June 2020...

**Comment by iMonZ** [comment]
> Stating the obvious - the large batch of linked reports in the middle of 2020 is likely because people were upgrading to Mac OS X 10.15, which also removes the ability to run applications in 32-bit mode. 10.15 was apparently out in oct 2019, but people were probably delaying upgrading until Big Sir was announced in June 2020...

DotNet 6 promised full 64bit support for all library’s and frameworks… let’s see if this will be true

**Comment by filipnavara** [comment]
> DotNet 6 promised full 64bit support for all library’s and frameworks… let’s see if this will be true

WinForms are only officially support on Windows. That's not changing with .NET 6.

We will release unofficial packages of our Mono's WinForms fork on https://github.com/emclient/mac-playground for .NET 6. We're currently running it experimentally on the public previews and it seems promising (not all fixes are published to the repository yet). The general code works on both CoreCLR and MonoVM in .NET 6 Preview 5+.

**Comment by TV4Fun** [comment]
Hey, it's 2022, any update on this?

**Comment by hngdog** [comment]
seriously. running ckan in a terminal window sucks hard. I have brain fog when I look at trying to piece this together on my own. Anyone care to redo the mac version of this stuff? Macs are only 3 versions in on 64-bit only...

**Comment by yuppox** [comment]
> > DotNet 6 promised full 64bit support for all library’s and frameworks… let’s see if this will be true
> 
> WinForms are only officially supported on Windows. That's not changing with .NET 6.
> 
> We will release unofficial packages of our Mono's WinForms fork on https://github.com/emclient/mac-playground for .NET 6. We're currently running it experimentally on the public previews and it seems promising (not all fixes are published to the repository yet). The general code works on both CoreCLR and MonoVM in .NET 6 Preview 5+.

Any update on this? I installed Mono on my Mac but without Forms support, it feels rather pointless.

**Comment by iMonZ** [comment]
> > > DotNet 6 promised full 64bit support for all library’s and frameworks… let’s see if this will be true
> > 
> > 
> > WinForms are only officially supported on Windows. That's not changing with .NET 6.
> > We will release unofficial packages of our Mono's WinForms fork on https://github.com/emclient/mac-playground for .NET 6. We're currently running it experimentally on the public previews and it seems promising (not all fixes are published to the repository yet). The general code works on both CoreCLR and MonoVM in .NET 6 Preview 5+.
> 
> Any update on this? I installed Mono on my Mac but without Forms support, it feels rather pointless.

Forms is very outdated and legacy.
Try something like MAUI or AvaloniaUI.
They are plenty of tools that are modern and not old like WinForms.
How old are you? 60?

**Comment by TV4Fun** [comment]
> > > > DotNet 6 promised full 64bit support for all library’s and frameworks… let’s see if this will be true
> > > 
> > > 
> > > WinForms are only officially supported on Windows. That's not changing with .NET 6.
> > > We will release unofficial packages of our Mono's WinForms fork on https://github.com/emclient/mac-playground for .NET 6. We're currently running it experimentally on the public previews and it seems promising (not all fixes are published to the repository yet). The general code works on both CoreCLR and MonoVM in .NET 6 Preview 5+.
> > 
> > 
> > Any update on this? I installed Mono on my Mac but without Forms support, it feels rather pointless.
> 
> Forms is very outdated and legacy. Try something like MAUI or AvaloniaUI. They are plenty of tools that are modern and not old like WinForms. How old are you? 60?

Or possibly just wanting to run some legacy code and not wanting to port it to a new UI library? How old are you? 12?

**Comment by bkimmett** [comment]
Unfortunately, many people use WinForms because it is easy and quick to use. A variety of small tools for data extraction and processing from one file format or another are built in it.

So adding compatibility for it isn't going to be a problem that goes away any time soon.

**Comment by transmatrix** [comment]
Can't believe this has been an issue for 4 years and still isn't resolved, and it appears that no one is working on this as no one is assigned to the ticket.

**Comment by tudor-turcu** [comment]
@transmatrix - this is not simply an .. issue - but a huge amount of work to port a framework to a new, previously unsupported OS/platform.. Since it's not something sponsored by Microsoft, only the work of some volunteers, it will be a miracle if it will be ever completed.

**Comment by vadim-a-yegorov** [comment]
> no32exec

Have you tried this?

https://gist.github.com/Cadichon/1122d2398a1d6a4b410bd5e91d3b060d

**Comment by MaddTheSane** [comment]
That won't work on Apple Silicon Macs.

**Comment by vadim-a-yegorov** [comment]
> That won't work on Apple Silicon Macs.

Apple Silicon Macs are designed to watch series, no one in their right mind would buy it except as a gift for a girlfriend. So if you're developer targeting M1 – try javascript.

**Comment by MaddTheSane** [comment]
I think Apple has completely migrated from Intel this WWDC.

**Comment by vadim-a-yegorov** [comment]
That means we screwed completely.

**Comment by TV4Fun** [comment]
> Apple Silicon Macs are designed to watch series, no one in their right mind would buy it except as a gift for a girlfriend...

No, that is just flat out wrong. Intel's viability has been flagging for years and is especially poorly suited for laptops because the fundamental design of the architecture makes optimizing power usage basically impossible. ARM is the future and of you haven't tried doing dev work on an M1 laptop, I would encourage it as it's a wonderful experience.

**Comment by vadim-a-yegorov** [comment]
> you're not into series

I see you're quite into series, that's fine :)

Talking seriously all servers I run for my services and stuff are ARM-based and that's great, I can confirm as a low-level engineer, I love them. But not my personal workstation. It's too early for that, we should wait until 2027.

**Comment by TV4Fun** [comment]
I've been using an M1 MacBook Pro as my main workstation since 2021. I had a lot of problems with compatibility when I first got it, but by now pretty much every major library and dev tool has M1 compatibility except for some older ones no one really uses any more 😉.

**Comment by vadim-a-yegorov** [comment]
So is this Mono also rusty piece that nobody uses anymore? I'd rather not recommend to buy Silicon Mac.

**Comment by HebaruSan** [comment]
> Have you tried this?
> 
> To re-enable 32-bit processing from the kernel, just disable Apple's disabling function by adding no32exec=0 to your boot-args (`sudo nvram boot-args="no32exec=0"` if SIP is disabled, if not run the same command from Recovery mode).
> 
> To get the 32-bit libraries back, just copy them from Mojave and add `export DYLD_ROOT_PATH=...` before running the executable or add this line to your `~/.zshrc` file.

No, I haven't tried that. I'm not a Mac user but rather a contributor to a .NET app that others wish to run on Macs. A kernel config hack and copying libraries from a version of the OS that they may not even have access to is a non-starter for a broad userbase of varying tech expertise, I'm afraid.

I'd also ask some understanding that some folks may be subscribed to email notifications here in case there are updates on WinForms support on Macs. There are other places to debate hardware choices without spamming those users.

**Comment by Sappharad** [comment]
[filipnavara](https://github.com/filipnavara)'s port posted in earlier in the thread is still working for me on .NET 6.

It's a little bit more challenging to consume for ports because any dependencies will need to be patched and built against it as well, but I don't know why people are still arguing in this thread about it. They don't want people developing new apps against WinForms and the solution posted previously is better than the official Mono version ever was.

**Comment by chucker** [comment]
> But it's interesting to see how Mono devs are picking their nose rather than support M1

This is very off-topic, but most if not all of the Mono team now works on .NET instead. Leaving aside technical hurdles, there's simply not much manpower. And both Mono and WinForms are arguably legacy.

**Comment by tudor-turcu** [comment]
> So is this Mono also a rusty piece that nobody uses anymore? I don't think so, I'd rather not recommend to buy an Apple Silicon Mac. But it's interesting to see how Mono devs are picking their nose rather than support M1, it means they silently agree with me.

@vadim-a-yegorov - everybody knows that mono is no longer actively developed for more than 3 years; it's practically in 'support' mode, with just a few volunteers working on it. 
For all new development, .NET Core is the replacement technology, even if it's not completely backward-compatible with .NET Framework or mono, and some old technologies are no longer present in .NET Core.
Nowadays, mono is used mostly by old .NET Framework projects that for one reason or another have to run on other platforms than Windows (like Linux), but can't migrate to .NET Core (we are in this situation, with a large project).
I would be more than happy to be able to run out WinForms application on MacOSX without using wine, just mono, but I don't have too much hopes..

**Comment by migueldeicaza** [comment]
Your best bet, remains the one that was done a few years ago: use alternative winforms implementations that exist out there.

They are not perfect matches, but close enough.   Sadly, they require some assembly.  I tried a few years ago to merge the changes necessary

**Comment by vricosti** [comment]
So no news about this issue ?  Is [mac-playground](https://github.com/filipnavara/mac-playground) a replacement of mono with some improvements or something complementary ?

**Comment by filipnavara** [comment]
> Is [mac-playground](https://github.com/filipnavara/mac-playground) a replacement of mono with some improvements or something complementary ?

It's something we use at @emclient for delivering a macOS version of our app on a modern .NET (ie. version 8+). However, we didn't update the public repository with fixes for a long time. The general interest was lackluster and when `git subtree` broke due to a bug on our internal repository I stopped doing the code sync to the public one. Notably, years later the `git subtree` issue is still not fixed (the history of the repo triggers a pathologic behavior resulting in a stack overflow) and so syncing the contents is a non-trivial task...

**Comment by vricosti** [comment]
> > Is [mac-playground](https://github.com/filipnavara/mac-playground) a replacement of mono with some improvements or something complementary ?
> 
> It's something we use at [@emclient](https://github.com/emclient) for delivering a macOS version of our app on a modern .NET (ie. version 8+). However, we didn't update the public repository with fixes for a long time. The general interest was lackluster and when `git subtree` broke due to a bug on our internal repository I stopped doing the code sync to the public one. Notably, years later the `git subtree` issue is still not fixed (the history of the repo triggers a pathologic behavior resulting in a stack overflow) and so syncing the contents is a non-trivial task...

Ok thanks.
If I have a few days so I will try to take your work and see if I can make something with it...
[https://gitlab.winehq.org/vricosti/mono](https://gitlab.winehq.org/vricosti/mono) on branch dev/mac-playground

I am new to this stuff so when compiling there are two define XAMARINMAC and MONOMAC, which one is supposed to be used when compiling mono (I am following instructions and I enter make without any additional flags). 
From what I understand when I compile none of them are used.
I added the two folders System.Windows.Forms.CocoaInternal and System.Windows.Forms.Extensions and for the next step I think I will create a System.Drawing.Mac that will contains the source file from sysdrawing-coregraphics/System.Drawing because I am not sure I want to follow the way of libgdiplus as in Mac-playground/mono/mcs/class/System.Drawing (Except if you tell me it's better to use libgdiplus... but the less layer the better)

On second thought maybe the work you have done with mono/mcs/class/System.Drawing is better for integration in official mono repo, what do you think ?


Then I have some errors with the Cocoa types:

```
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(4,22): error CS0234: The type or namespace name 'Mac' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(4,22): error CS0234: The type or namespace name 'Mac' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(4,22): error CS0234: The type or namespace name 'Mac' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(8,7): error CS0246: The type or namespace name 'MacApi' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/Extensions.cs(3,7): error CS0246: The type or namespace name 'MacApi' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/IMacNativeControl.cs(5,7): error CS0246: The type or namespace name 'MonoMac' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/MonoView.cs(29,22): error CS0234: The type or namespace name 'Mac' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)
System.Windows.Forms.CocoaInternal/Swizzle.cs(3,7): error CS0246: The type or namespace name 'MacApi' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/Troubleshooter.cs(3,7): error CS0246: The type or namespace name 'ObjCRuntime' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/WindowsEventResponder.cs(2,22): error CS0234: The type or namespace name 'Mac' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)
System.Windows.Forms.CocoaInternal/XplatUICocoa.Dnd.cs(8,22): error CS0234: The type or namespace name 'Mac' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)
System.Windows.Forms.CocoaInternal/XplatUICocoa.cs(61,22): error CS0234: The type or namespace name 'Mac' does not exist in the namespace 'System.Drawing' (are you missing an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(127,42): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(89,39): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(151,23): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(159,26): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(168,28): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(177,27): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(187,33): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(20,3): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectPasteboard.cs(22,31): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(22,38): error CS0246: The type or namespace name 'NSPasteboardItemDataProvider' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(84,43): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(84,64): error CS0246: The type or namespace name 'NSPasteboardItem' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(113,49): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(131,30): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(131,51): error CS0246: The type or namespace name 'NSPasteboardItem' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(138,30): error CS0246: The type or namespace name 'NSPasteboard' could not be found (are you missing a using directive or an assembly reference?)
System.Windows.Forms.CocoaInternal/DataObjectProvider.cs(138,51): error CS0246: The type or namespace name 'NSPasteboardItem' could not be found (are you missing a using directive or an assembly reference?)
```

**Comment by vricosti** [comment]
I am posting here in case a developer will continue my work since my health is not that good:

### Build last mono sources (needs to have an existing mono install)

Install mono binaries: [https://www.mono-project.com/download/stable/#download-mac](https://www.mono-project.com/download/stable/#download-mac)
Then check it's in your path in my case I had to add it to my shell init (.zshrc):

```
export MONO_HOME=/Library/Frameworks/Mono.framework/Home
export PATH=$MONO_HOME/bin:$PATH
```

Check it works:

```
~ /Library/Frameworks/Mono.framework/Home/bin/mono --version
Mono JIT compiler version 6.12.0.206 (2020-02/0cbf0e290c3 Tue Feb 13 17:04:08 EST 2024)
...
```

Now we have a working mono install let's build mono from sources:  

```
brew install autoconf automake libtool pkg-config cmake python3
mkdir mono-mac-playground
cd mono-mac-playground
git clone --recurse-submodules -b dev/mac-playground  git@gitlab.winehq.org:vricosti/mono.git
cd mono
export PREFIX=/usr/local
./autogen.sh --prefix=$PREFIX --disable-nls
make
sudo make install
```

Once it's done ,check our fresh build of mono is working and update your path (now pointing to /usr/local):

```
export MONO_HOME=/usr/local
export PATH=$MONO_HOME/bin:$PATH
```

```
source ~/.zshrc 
~ mono --version
Mono JIT compiler version 6.13.0 (main/04ae7e04a1c Jeu  3 avr 2025 12:21:22 UTC)
...
```

### Build macios (seems to be the new Xamarin.Mac)

They have renamed Xamarin. into Microsoft.:  

* `Xamarin.iOS.dll` -> `Microsoft.iOS.dll`
* `Xamarin.TVOS.dll` -> `Microsoft.tvOS.dll`
* `Xamarin.MacCatalyst.dll` -> `Microsoft.MacCatalyst.dll`
* `Xamarin.Mac.dll` -> `Microsoft.macOS.dll`


You need to have at least Xcode 16.3. 

```
git clone --recurse-submodules -b dotnet-9.0.1xx-xcode16.3-9288 https://github.com/dotnet/macios.git
cd macios
brew install shellcheck yamllint p7zip 
```

Inside make.config modify the line XCODE_DEVELOPER_ROOT=...
`XCODE_DEVELOPER_ROOT=/Applications/Xcode.app/Contents/Developer`

call system-dependencies.sh it will make you install some pkg and once it's ok call make world:  
  
```
./system-dependencies.sh 
make world
```

**Comment by filipnavara** [comment]
You just cannot put the mac-playground implementation back into classic Mono because it has an inherent dependency on the Xamarin macOS bindings. It simply cannot work without some form of Cocoa bindings and the Xamarin bindings (now no longer supported on classic Mono) require a lot of custom loader code so it won't work on unmodified apps.

**Comment by vricosti** [comment]
> You just cannot put the mac-playground implementation back into classic Mono because it has an inherent dependency on the Xamarin macOS bindings. It simply cannot work without some form of Cocoa bindings and the Xamarin bindings (now no longer supported on classic Mono) require a lot of custom loader code so it won't work on unmodified apps.

You mean macios (the new Xamarin.Mac does not support mono ... ?)

**Comment by filipnavara** [comment]
> You mean macios (the new Xamarin.Mac does not support mono ... ?)

No. The legacy Mono is for intents and purposes not supported anymore, only MonoVM (part of .NET 6+).

**Comment by t-turcu-qdi** [comment]
> No. The legacy Mono is for intents and purposes not supported anymore, only MonoVM (part of .NET 6+).

That's true, however mono was practically abandoned in the last 4-5 years, so nobody was expecting any kind of technical support from any company. :-) 
However, there are many large projects still using it in production, that have not managed yet to migrate and rewrite the code for .NET Core.

**Comment by vricosti** [comment]
Ok so I can give up...

**Comment by madewokherd** [comment]
> You just cannot put the mac-playground implementation back into classic Mono because it has an inherent dependency on the Xamarin macOS bindings. It simply cannot work without some form of Cocoa bindings and the Xamarin bindings (now no longer supported on classic Mono) require a lot of custom loader code so it won't work on unmodified apps.

Is it not possible to backport the updated bindings?

I'd be interested in integrating this, but at the moment I'm clueless about the mac-specific stuff.

**Comment by madewokherd** [comment]
FWIW, most of the current mono-based winforms development is happening at https://github.com/DanielVanNoord/System.Windows.Forms, and since they're targeting modern .NET it should be possible to use the current bindings. Maybe they'd be interested merging it.
